rclpy/src/rclpy/qos.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/rcl.h>
+#include <rcl/time.h>
+#include <rmw/types.h>
+
+#include <cstring>
+#include <memory>
+#include <string>
+
+#include "rclpy_common/common.h"
+#include "rclpy_common/handle.h"
+
+#include "qos.hpp"
+
+namespace rclpy
+{
+
+namespace
+{
+
+// Fill a given rmw_time_t from an rclpy.duration.Duration instance.
+void
+_convert_py_duration_to_rmw_time(py::capsule pyduration, rmw_time_t * out_time)
+{
+  assert(out_time != nullptr);
+  auto duration = static_cast<rcl_duration_t *>(pyduration);
+  out_time->sec = RCL_NS_TO_S(duration->nanoseconds);
+  out_time->nsec = duration->nanoseconds % (1000LL * 1000LL * 1000LL);
+}
+
+}  // namespace
+
+py::capsule
+convert_from_py_qos_policy(
+  int qos_history,
+  int qos_depth,
+  int qos_reliability,
+  int qos_durability,
+  py::capsule pyqos_lifespan,
+  py::capsule pyqos_deadline,
+  int qos_liveliness,
+  py::capsule pyqos_liveliness_lease_duration,
+  bool avoid_ros_namespace_conventions)
+{
+  auto qos_profile =
+    std::unique_ptr<rmw_qos_profile_t, decltype(&PyMem_Free)>(
+    static_cast<rmw_qos_profile_t *>(
+      PyMem_Malloc(sizeof(rmw_qos_profile_t))), PyMem_Free);
+  if (!qos_profile) {
+    throw std::bad_alloc();
```

Comment:
I believe this isn't documented in the header.

Reply:
Good point. See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.